library/L2_HAL/sensors/battery/ltc4150.hpp

```diff
@@ -4,95 +4,74 @@
-    /// 3600 is X 32.55 is kGvF and 0.00005 is kRsense
-    static float GetBatterymAh()
-    {
-      return static_cast<float>(static_cast<float>(battery_info.num_ticks) /
-                                (3600 * kGvf * kRsense));
-    }
-  };
-
-  /// @param isr - handler for coulomb counter ticks.
   /// @param int_pin - gpio pin for handling handling interrupts from the
-  /// LTC4150
+  ///        LTC4150
   /// @param pol - gpio pin to determine the polarity output of the LTC4150.
-  template <int id>
-  explicit constexpr Ltc4150(const TickHandler<id> & isr,
-                             Gpio & int_pin,
-                             Gpio & pol)
+  /// @param resistance - value of impedance represented in ohms for
+  ///        calculating battery charge
+  explicit constexpr Ltc4150(sjsu::Gpio & int_pin,
+                             sjsu::Gpio & pol,
+                             units::impedance::ohm_t resistance)
       : int_pin_(int_pin),
-        pol_(pol),
-        get_mah_(isr.GetBatterymAh),
-        isr_handler_(isr.IsrHandler)
+        pol_pin_(pol),
+        resistance_(resistance),
+        pulses_(0),
+        polarity_(Polarity::kDischarging)
   {
   }
 
   /// Initialize hardware, setting pins as inputs and attaching ISR handlers to
   /// the interrupt pin.
-  void Initialize() const override
+  void Initialize() override
   {
-    pol_.SetAsInput();
+    pol_pin_.SetAsInput();
     int_pin_.SetAsInput();
-    int_pin_.AttachInterrupt(isr_handler_, Gpio::Edge::kEdgeFalling);
+    pol_pin_.AttachInterrupt(
+        [this]() {
+          polarity_ =
+              (pol_pin_.Read()) ? Polarity::kDischarging : Polarity::kCharging;
+        },
+        Gpio::Edge::kEdgeFalling);
+    int_pin_.AttachInterrupt([this]() { pulses_++; }, Gpio::Edge::kEdgeFalling);
   }
 
-  /// @return the calculated mAh from the TickHandler instance
-  float GetBatterymAh() const override
+  /// @return the calculated mAh
+  units::charge::milliampere_hour_t GetCharge() const override
   {
-    return get_mah_();
+    /// 3600 is X 32.55 is kGvF and 0.00005 is kRsense
+    auto sign = (polarity_ == Polarity::kCharging)
+                    ? units::dimensionless::scalar_t(-1)
+                    : units::dimensionless::scalar_t(1);
+    /// We cast the pulses to scalar so we can calculate mAh
+    float pulses = static_cast<float>(pulses_);
+    return sign * units::charge::milliampere_hour_t{
+      pulses / (3600.0f * kGvf.to<float>() * resistance_.to<float>())
+    };
   }
 
  private:
-  /// Gpio pin to recieve tick interrupts from the LTC4150.
   sjsu::Gpio & int_pin_;
-  /// Gpio pin to check polarity output from the LTC4150.
-  sjsu::Gpio & pol_;
-  /// function pointer to grab current battery mAh
-  float (*get_mah_)(void) = nullptr;
-  /// function pointer to isr handler
-  void (*isr_handler_)(void) = nullptr;
+  sjsu::Gpio & pol_pin_;
+  units::impedance::ohm_t resistance_;
+  int pulses_;
+  Polarity polarity_;
```

Comment:
```suggestion
    std::atomic<Polarity> polarity_;
```
Use atomic here to make it clear that a set, read or get operation on this field will occur atomically, since ISR have the capability to nest themselves.